adr/query_deduplication.tla

```diff
@@ -0,0 +1,501 @@
+------------------------ MODULE query_deduplication ------------------------
+
+
+EXTENDS Integers, Sequences, FiniteSets, TLC
+
+CONSTANT Proc
+CONSTANT Keys
+(*
+
+\* from https://probablydance.com/2020/10/31/using-tla-in-the-real-world-to-understand-a-glibc-bug/
+--algorithm Spinlock 
+{
+  variables
+    cache_lock = 0;
+    wait_map_lock = 1;
+    locks = [i \in {cache_lock, wait_map_lock} |-> FALSE];
+    \* set to false for every possible key at first
+    cache = [ i \in Keys |-> FALSE ];
+    wait_map = [ i \in Keys |-> {} ];
```

Comment:
the wait map is a hashmap  of `key -> set of processes`. Sets are represented by curly braces. Each process is a task that runs concurrently with the other processes